CARDS-1262: The suggestions dropdown for Vocabulary Questions should display additional fields as needed when the suggestion label doesn't match the user input

veronikaslc opened the pull request.

Commit:
CARDS-1262: Display synonyms in the suggestions dropdown for Vocabulary Questions

sdumitriu reviewed (changes requested):
This should also work for the other synonym field, `synonym`, which is what HPO uses.

Commit:
CARDS-1262: Display synonyms in the suggestions dropdown for Vocabulary Questions
Added synonym HPO field into account

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -209,7 +209,7 @@ function VocabularyQuery(props) {
         suggestions.push(
           <MenuItem
-            className={classes.dropdownItem}
+            className={classes.dropdownItem + " " + (element["has_exact_synonym"] ? classes.dropdownHasSynonymItem : '')}
```
Only `has_exact_synonym` is checked here, while when displaying we also look at `synonym`.

marta- replied:
Addressed in b97a710

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -235,6 +235,23 @@ function VocabularyQuery(props) {
+          let synonyms = element["has_exact_synonym"] || element["synonym"] || null;
+          if (synonyms && synonyms.length > 0) {
+            suggestions.push(
+              <MenuItem
+                className={classes.dropdownSynonymItem}
+                key={element["has_exact_synonym"][0]}
+              >
+                <Typography
+                  component="p"
+                  variant="caption"
+                  color="textSecondary"
+                >
+                  {"Also known as: " + synonyms.join(", ")}
+                </Typography>
+              </MenuItem>
```
With the synonym as a separate `MenuItem`, I find the interaction uncomfortable: hovering the synonym line produces a separate shading, and clicking on it doesn't do anything. The whole group (name + synonyms) should work as a whole.
![CPT2107261525-787x544](https://user-images.githubusercontent.com/651980/127046254-384bd730-ba06-4e7e-9a77-c4254c197b7c.gif)

There's also a layout issue with the answers added from suggestions (briefly visible at the end of the screencast above), but it appears to be unrelated to these changes and will be reported separately.

marta- replied:
Addressed in b97a710

marta- replied:
> There's also a layout issue with the answers added from suggestions (briefly visible at the end of the screencast above), but it appears to be unrelated to these changes and will be reported separately.

Reported as https://phenotips.atlassian.net/browse/CARDS-1279

marta- committed:
CARDS-1262: Display synonyms in the suggestions dropdown for Vocabulary Questions
UX improvement: the synonyms should be in the same MenuItem as the name
Misc bugfix: for HP, display the synonym in the info box

marta- committed:
CARDS-1262: Display synonyms in the suggestions dropdown for Vocabulary Questions
Remove "Also known as" and display synonyms one per line to improve readability

marta- commented:
The last commit 4c7b1b1 removes the text "Aslo known as" and dsplays synonyms one per line to make them easier to read. The readability problem was quite prominent for HP, where synonyms can be quite long and some contain commas. This makes the implementation diverge from the original specs, and feedback on the change is welcome.

Question: should we limit the number of synonyms being displayed? In HP the list can be quite long. This is an example of what the suggestion list might look like:

![image](https://user-images.githubusercontent.com/651980/127058000-8bbfc4c0-cf34-40e1-9b9c-66aec37670b7.png)

veronikaslc commented:
I'd say we should limit the number of synonyms just for a display purpose.

marta- committed:
CARDS-1262: Display synonyms in the suggestions dropdown for Vocabulary Questions
UX improvement - refined the conditions for displaying synonyms in the suggestions dropdown:
* if the user input matches the suggestion label, don't display anything else
* if the user input does not match the suggestion label, display only enough synonyms matching the entire user's query
* if there are no matching synonyms, display a fragment of the definition matching the first word of the query
Misc: fixed js console warning introduced in previous commit (missing key prop)

marta- commented:
The last commit 6037f8e brings even more divergence from the original specs:
* if the user input matches the suggestion label, don't display anything else
* if the user input does not match the suggestion label, display only enough synonyms matching the entire user's query
* if there are no matching synonyms, display a fragment of the definition matching the first word of the query

The goal is to show the user why we're providing each specific suggestion (how it matches their query) without taking up too much screen space with irrelevant details. See two examples in the screenshots below. Feedback is welcome.
![image (46)](https://user-images.githubusercontent.com/651980/127220763-fcff4b64-1245-4405-8242-f687f94c8e7a.png)
![image (47)](https://user-images.githubusercontent.com/651980/127220766-90b57485-1a77-4aea-9f65-96d98fc1358d.png)

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -204,9 +205,25 @@ function VocabularyQuery(props) {
     if (data["rows"]?.length > 0) {
       data["rows"].forEach((element) => {
         var name = element["label"] || element["name"] || element["identifier"];
+        var synonyms = element["synonym"] || element["has_exact_synonym"] || [];
+        var definition = Array.from(element["def"] || element["description"] || element["definition"] || [])[0] || "";
```
We duplicate some of the code for retrieving synonyms and definitions here (similar code is in VocabularyBrowser). Ideally we should move the logic for identifying various fields out of the UI component and have a unified way of accessing the data, that is robust to the different field names used by different vocabularies. But this optimization is out of scope for now.

marta- committed:
CARDS-1262: The suggestions dropdown for Vocabulary Questions should display additional fields as needed when the suggestion label doesn't match the user input
Fixed regression: cannot add a suggestion when clicking on the text

sdumitriu reviewed (commented):
Clicking on a synonym doesn't select the term, just on the label or all the whitespace.

Would it work better to stop the event when clicking on the info button and not add any more checks before accepting the suggestion?

marta- committed:
CARDS-1262: The suggestions dropdown for Vocabulary Questions should display additional fields as needed when the suggestion label doesn't match the user input
Fixed UX issue: clicking on tne synonym now adds the suggestion

sashaandjic commented:
Is this acceptable?
![image](https://user-images.githubusercontent.com/23243244/127270394-262ea2ec-61e6-4fb3-a914-de55b4d19c24.png)

sashaandjic commented:
We should prevent this situations as well. 
![image](https://user-images.githubusercontent.com/23243244/127271375-3229e7da-1577-4639-b7be-7a722f54df61.png)

marta- commented:
> Suggestion: if there is a 100% match for the synonym, do not offer free text as an option.

Done in 6db5ca2 .

marta- committed:
CARDS-1262: The suggestions dropdown for Vocabulary Questions should display additional fields as needed when the suggestion label doesn't match the user input
Don't show the user entry in the dropdown as an option when it matches a synonym exactly
Also, ignore case when comparing the user entry with both name and synonyms

sashaandjic reviewed (approved):
tested
